### ISSUE TITLE
Inline script loader seems to conflict with good Content Security Policy practices

### ISSUE THREAD

**Issue by HenrikJoreteg** [parcel-bundler/parcel#6841]
# 🙋 feature request

Using dynamic `import()` generates inline JS, which means you cannot use Parcel 2 along with a good Content Security Policy that disallows inline script.

here's the relevant code:

https://github.com/parcel-bundler/parcel/blob/6719204a36a378506908f862a6b70e24cbfaa1fc/packages/runtimes/js/src/helpers/browser/import-polyfill.js#L19-L30

## 🤔 Expected Behavior

I would like to be able to use:

```js
import('../my-other-module')
``` 

And still have a CSP like this: 

```html
<meta http-equiv="Content-Security-Policy" content="script-src 'self' https://polyfill.io>
```

## 😯 Current Behavior

In order for the browser to still load my bundle-split code, I have to change my CSP to add `unsafe-inline` which... unfortunately somewhat negates the point of using a CSP to prevent XSS.

## 💁 Possible Solution

If we could either avoid using an inline script as a way to load script. Or if we could somehow generate a content hash for the script to appease the CSP, that might work as well. There are some details on MDN about that here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#:~:text=%27%3Chash-algorithm%3E-%3Cbase64,for%20external%20scripts.

## 🔦 Context

1. I am building a security sensitive medical application that handles Protected Health Information (PHI) and therefore is covered under [HIPAA](https://www.hhs.gov/hipaa/index.html).
2. I recently updated this application from Parcel 1 to Parcel 2. Parcel 1 did not have this issue, Parcel 2 does.
3. The mechanism Parcel 2 uses to lazy-load modules uses inline JS
4. I have a Content Security Policy in place, but in order to use Parcel 2, I have to add `unsafe-inline` which seems like a poor solution.

## 💻 Examples

Here's what happens after updating to Parcel 2, if I don't add the `unsafe-inline` directive: 

<img width="601" alt="Screen Shot 2021-08-30 at 4 55 06 PM" src="https://user-images.githubusercontent.com/67925/131420381-b4df4aa0-8c6e-4a25-b086-ef173bf1e7de.png">

I really, really love Parcel, I'm super excited about the speed boost I'm seeing w/ Parcel 2! Thanks y'all for this incredible alternative to the configuration headaches I've suffered through w/ webpack in the past :) <3

**Comment by devongovett** [comment]
Hmm are blob URLs allowed? Sounded like maybe not in some browsers from MDN...

Reason for this existing is that technically browsers that support dynamic import are a subset of those that support module scripts.

**Comment by HenrikJoreteg** [comment]
Hmm... the whole dynamic import being a subset thing... It seems like there's a really small gap there, no?

https://caniuse.com/?search=JavaScript%20modul

https://caniuse.com/?search=JavaScript%20modules

**Comment by devongovett** [comment]
Yeah. If you've told Parcel via a browserslist that you support older browsers that don't support module  scripts, we generate a nomodule version. Then if we used dynamic import in the module version, the site will be broken in browsers that support modules but not dynamic import because it won't load the fallback. Would that be strange? Feels weird to have a hole in browser support there.

Note that if your browserslist is setup to only include browsers supporting dynamic import then it is used natively without the polyfill.

**Comment by devongovett** [comment]
What's your browserslist query? Maybe there's something wrong in our logic...

**Comment by HenrikJoreteg** [comment]
@devongovett hm... interesting. this is what I had (I can probably safely update the version of safari at this point). 

```
"browserslist": [
    "defaults",
    "safari 11",
    "not IE 11"
  ],
```

But... does it use the polyfill if any of the browsers would have required it maybe? Obviously, Chrome (as per that screenshot) shouldn't need it, but is using it.

**Comment by devongovett** [comment]
Ah yeah it's probably Safari 11. Looks like dynamic import is supported in 11.3 and later. Maybe better to go with a usage percentage instead? I think the "defaults" setting would already do this.

**Comment by HenrikJoreteg** [comment]
I tried a few different combinations of things, surprisingly many of them ended up with a nomodule version... so finally just went for the direct approach:

`"supports es6-module-dynamic-import and > .5%"`

Which, worked beautifully, and the resulting list of browsers is fine for my audience.

I really appreciate your help!

I'm set at this point, but do you think this is a bug? I mean... the thing I personally wasn't expecting was that Chrome was using the polyfill code. Not sure whether to consider this an issue or not.

**Comment by devongovett** [comment]
Great! I'll double check the logic again in more detail and see if something is broken.

**Comment by devongovett** [comment]
Ah, I guess this is how webpack deals with this issue. Server sets `__webpack_nonce__`, which is used as the `nonce` attribute on inline scripts and set in the CSP. https://webpack.js.org/guides/csp/

**Comment by HenrikJoreteg** [comment]
Makes sense 👍

**Comment by jamesmcallister** [comment]
Having the same issue this week with safari, the browser list change helped, thanks @HenrikJoreteg , just need to spend time checking we can still support our customers browsers.

On the CSP nonce. have you seen how google use nonces with tag manager? https://developers.google.com/tag-manager/web/csp

Seems to be a better approach to the webpack way, as nonce will be inserted in to the script tag from the server request.
